File: LC/[2762] Continuous Subarrays.cpp
using LL=long long;
class Solution {
public:
    long long continuousSubarrays(vector<int>& nums) {
        /*
        * subarray (continuous): abs(diff of any pair) <= 2 ===> (max-min) <=2
        * if we have a window (ll, rr) were (max-min) <= 2
        * -> number of continuous array which ends on rr -> rr-ll+1
        * ll=rr=0
        * expand rr and accumulate count of continuous array (ends on rr)
        * when the subarray[ll:rr] disqualified, increment ll until subarray[ll:rr] qualify again   

        * to maintain sliding window max/min -> deque
        * Time: O(N)
        * Space:O(N)
        */
        deque<int> dq_max, dq_min;
        int ll=0, rr=0, N=nums.size();
        LL ans=0;
        for(rr=0; rr < N; ++rr ){//for each subarray ending with rr
            //find min/max in the subarray
            //shrink the subarray until (max-min) <=2
            //count, accumulate the number      
            while(!dq_max.empty() && nums[rr] > nums[dq_max.back()]) dq_max.pop_back();
            dq_max.push_back(rr);
            
            while(!dq_min.empty() && nums[rr] < nums[dq_min.back()]) dq_min.pop_back();
            dq_min.push_back(rr);

            while(!dq_max.empty() && !dq_min.empty() && (nums[dq_max.front()] - nums[dq_min.front()]) > 2){
         
                while(!dq_max.empty() && dq_max.front() <= ll) dq_max.pop_front();
                while(!dq_min.empty() && dq_min.front() <= ll) dq_min.pop_front();
                ++ll;
            }  
            ans += (rr-ll+1);
        }

     return ans;
    }
};

File: LC/[2762] Continuous Subarrays.py
from collections import deque
class Solution:
    def continuousSubarrays(self, nums: List[int]) -> int:
        ans=0
        dq_max=deque() # maintain max in the sliding window
        dq_min=deque() # maintain min in the sliding window

        # sliding window boundary
        ll, rr, N = 0, 0, len(nums)

        # for each window right-end rr, calc the qualified continuous array
        for rr in range(N):
            # maintain window max/min
            while len(dq_max) > 0 and nums[rr] > nums[dq_max[-1]]: dq_max.pop()
            dq_max.append(rr)
            
            while len(dq_min) > 0 and nums[rr] < nums[dq_min[-1]]: dq_min.pop()
            dq_min.append(rr)

            # adjust sliding window to be qualifed continuous array
            while len(dq_max) > 0 and len(dq_min) > 0 and (nums[dq_max[0]]-nums[dq_min[0]]) > 2:
                while len(dq_max)> 0 and ll >= dq_max[0]: dq_max.popleft()
                while len(dq_min)> 0 and ll >= dq_min[0]: dq_min.popleft()
                ll+=1
            
            ans += (rr-ll+1)

        return ans